Refactoring of discovery for hosts and datastores

kremers opened the pull request:
- Fix errors with more complicated folder structure
- Refactor static logging to fine tunable logging

Folder structs like the following will work now:

Folder (Root)
- Folder (Company)
  - Folder (Country)
    - Datacenter Folder
      - Project Folder (Compute Ressource or CCR here)
        - Customer related Folder
          - Production/Dev Folder (Compute Ressource or CCR here)

Signed-off-by: Martin Kremers <info@martinkremers.de>

kremers commented:
Python != 3 unit test failed, but the exporter does not work anyways with Python2. So I vote for removal of this check.

pryorda commented:
Looking over this now.  As far as python 3 vs python 2. I prefer only look at python3.

pryorda commented:
https://travis-ci.org/pryorda/vmware_exporter/jobs/531910714 shows that the assertions are failing to match and there is somthing not right about StopIteration:?

kremers commented:
I'll have a look.
BTW I've created a helm chart for the app: https://github.com/helm/charts/pull/13719

pryorda commented on vmware_exporter/vmware_exporter.py:
```diff
@@ -369,58 +369,66 @@ def datacenter_inventory(self):
+        def _collect(node, level=1, dc=None, folder=None):
+            inventory = {}
+            if isinstance(node, vim.Folder) and not isinstance(node, vim.StoragePod):
+                logging.debug(f"[F ] {('-' * level).ljust(7)} {node.name}")
```
Don't Hate but for the [F ], CR H, DC. Can you add something a bit more descriptive? I hate having to bust out decoder rings.

kremers replied:
Changed this in the most up to date version and made flake8 happy.

I'll have to look into travis for the failing build. Trying to simulate it locally

pryorda commented on vmware_exporter/vmware_exporter.py:
```diff
@@ -873,27 +891,26 @@ class HealthzResource(Resource):
                         default=None, help="configuration file")
     parser.add_argument('-p', '--port', dest='port', type=int,
                         default=9272, help="HTTP port to expose metrics")
+    parser.add_argument('-l', '--loglevel', dest='loglevel',
```
I might of missed this, but can you do a env variable for this as well?

kremers replied:
Preferably in another change/PR, okay?

pryorda replied:
Works for me.

Jc2k commented:
@pryorda  Do we technically support python 3.5 (Ubuntu 18.04 LTS / Debian Stretch are on 3.5) right now do you know? I know we push people towards the container but iirc there are some not using it, and this PR will definitely introduce 3.6+ only features (f-strings). Either way, we should document minimum python version somewhere. I don't even think we say python3 only anywhere.

Jc2k commented:
We should probably add tests for a few different folder structures as this has repeatedly caught us out.

pryorda commented:
@Jc2k Yes we need to update the readme and since this is breaking we have to make sure the version gets bumped correctly. @kremers can you update the README with your changes too?

kremers commented:
> @pryorda Do we technically support python 3.5 (Ubuntu 18.04 LTS / Debian Stretch are on 3.5) right now do you know? I know we push people towards the container but iirc there are some not using it, and this PR will definitely introduce 3.6+ only features (f-strings). Either way, we should document minimum python version somewhere. I don't even think we say python3 only anywhere.

Since the Dockerfile is already based on 3.6 I doubt that we'll get into trouble.
I totaly agree regarding a comment about the minimal Python version! I failed trying it out with python 2.7 at first, just figuring out only python3 is supported.

Maybe the travis lint for the python2 can be removed to add some python 3.7 validation?

Jc2k commented:
It's the non Docker people i'm worried about :-)

Where's the python 2 lint hiding? I'm only seeing real python3 errors from travis I think?

I'll try and review this properly tomorrow AM - AFK this evening.

kremers commented:
locally pytest runs trough

`pytest --cov=vmware_exporter tests/unit
=========================================================================================== test session starts ============================================================================================
platform darwin -- Python 3.7.3, pytest-4.5.0, py-1.8.0, pluggy-0.11.0
rootdir: /Users/martinkremers/vmware_exporter
plugins: twisted-1.10, cov-2.7.1
collected 21 items

tests/unit/test_helpers.py ...                                                                                                                                                                       [ 14%]
tests/unit/test_vmware_exporter.py ..................                                                                                                                                                
...

======================================================================================== 21 passed in 0.49 seconds =========================================================================================
`

kremers commented:
I think one reason why there are failing tests, is that the mock objects are lacking the vim.* types.
I'll try to fix that.

kremers commented:
pls review 👍 

- Fixed tests
- Linter is happy now
- Changed debug logging to comply to 
- Fixed the names in the travis build and switched to matrix builds
- Extended the travis builds to contain python3.6 and python3.7

Jc2k commented:
Looking good. Thank you for your work on this.

A minor minior nit pick for future - it's easier for reviewers if you push your changes rather than squashing and force pushing. We can squash merge when the PR is merged to keep the history clean. This is more painful when there are a bunch of unrelated changes in the same PR.

Jc2k commented on .travis.yml:
```diff
@@ -1,40 +1,56 @@
+
+after_success:
+  - codecov
+  - if [ $INTEGRATION_TEST == "true" ]; then
```
Feels a bit weird to hang the release pipeline off `INTEGRATION_TEST` - should there be a different variable for the release pipeline?

Or should we have stages and have a release stage that only runs after the test stage?

kremers replied:
That is basically the same logic that has been put in the origin .travis.yml file. Only other step that was just called "python" was docker + integration tests. I just made it clearer by naming it this way.

Are you available in IRC? Maybe a short chat can clarify why I put things this way

Jc2k replied:
There is a vmware_exporter gitter.im channel that all the reviewers are on.

I am happy with the direction this is going in it was just an idea for making it even better. I can do a follow up PR if you aren't sure about stages.

Jc2k commented on tests/unit/test_vmware_exporter.py:
```diff
@@ -576,12 +577,20 @@ def test_vmware_get_inventory():
     datastore_2_folder.__dict__['name'] = 'datastore2-folder'
 
-    data_center_1 = mock.Mock()
+    data_center_1 = mock.Mock(spec=vim.Datacenter)
```
Can we have a few more variants of this test with different folder structures? With different levels of nesting.

kremers replied:
Yes! Would be a good idea! I'd like to add improved testing at least for my cases in a different PR.

Jc2k commented on vmware_exporter/vmware_exporter.py:
```diff
@@ -15,6 +15,7 @@
 import traceback
 import pytz
+import logging
 
```
I think it's more idiomatic to create a named logger rather than using the root one?

```
logger = logging.getLogger(__name__)
```

then `logger.info` instead of `logging.info` etc.

kremers replied:
Naming is hard, so I sticked to the root package for the application.
Someones might just prefer log over logger other than logging.

Jc2k commented on vmware_exporter/vmware_exporter.py:
```diff
@@ -489,47 +498,50 @@ def _vmware_get_datastores(self, ds_metrics):
-            ds_freespace = float(datastore.get('summary.freeSpace', 0))
-            ds_uncommitted = float(datastore.get('summary.uncommitted', 0))
-            ds_provisioned = ds_capacity - ds_freespace + ds_uncommitted
+            try:
```
Is this try just do catch where `datastore_labels` is missing a key or where `datastore` is missing a key? To avoid the nesting, would it be enough to do:

```python
try:
    name = datastore['name']
    labels = datastore_labels[name]
except KeyError:
    logging.exception(f'Datastore label for {name} missing')
    continue
```

kremers replied:
Implemented in the most up to date version

Jc2k commented on vmware_exporter/vmware_exporter.py:
```diff
@@ -662,79 +674,83 @@ def _vmware_get_vms(self, metrics):
 
         for host_id, host in results.items():
-            labels = host_labels[host_id]
+            try:
```
As above

kremers replied:
Implemented in the most up to date version

Jc2k commented:
@pryorda we should go with this approach instead of #104 - the approach there was unfinished and it switched to using the batch_fetch_properties for folder data which is slower in my tests.

pryorda commented:
I agree, let me see what you said in the rest. I'd like to see this get resolved asap haha.

kremers commented:
> Looking good. Thank you for your work on this.
> 
> A minor minor nit pick for future - it's easier for reviewers if you push your changes rather than squashing and force pushing. We can squash merge when the PR is merged to keep the history clean. This is more painful when there are a bunch of unrelated changes in the same PR.

Regarding the kind of contributions, some projects only allow full, cherry-pickable features to be merged. Squashing will possibly kill contribution information which will possibly hold people back from contributing. So you'll stick to merge all the commits, messing with the history or shadow the contribution under your name. For me it is easier to edit just a single commit rather then having a bunch of WIP commits that are rather unconnected.

Jc2k commented:
> > Looking good. Thank you for your work on this.
> > A minor minor nit pick for future - it's easier for reviewers if you push your changes rather than squashing and force pushing. We can squash merge when the PR is merged to keep the history clean. This is more painful when there are a bunch of unrelated changes in the same PR.
> 
> Regarding the kind of contributions, some projects only allow full, cherry-pickable features to be merged. Squashing will possibly kill contribution information which will possibly hold people back from contributing. So you'll stick to merge all the commits, messing with the history or shadow the contribution under your name. For me it is easier to edit just a single commit rather then having a bunch of WIP commits that are rather unconnected.

Yep, its all about trade offs as ever. I always say its not about making the /right/ decision or change, its about deciding who or what gets the pain. There is always pain. It's definitely easier to do it your way when your PR's don't have multiple changes in.

pryorda commented:
#66 #63 should be fixed by this
